File: 1001-2000/1866-number-of-ways-to-rearrange-sticks-with-k-sticks-visible/1866-number-of-ways-to-rearrange-sticks-with-k-sticks-visible.cpp
// Precompute 2D-DP
class Solution {
  constexpr static int M = 1000;  // maximum n
  using Matrix = array<array<int64_t, M + 1>, M + 1>;

  constexpr static int64_t modulo = 1e9 + 7;
  constexpr static int64_t mod(int64_t x) { return x % modulo; }

  const static Matrix init() {
    Matrix dp = {};
    dp[0][0] = 1;
    for (int n = 1; n <= M; ++n) {
      for (int k = 1; k <= min(n, M); ++k) {
        dp[n][k] = mod(dp[n - 1][k - 1] + (n - 1) * dp[n - 1][k]);
      }
    }

    return dp;
  }
  inline const static Matrix dp = init();

 public:
  int rearrangeSticks(int n, int k) { return dp[n][k]; }
};

File: 1001-2000/1866-number-of-ways-to-rearrange-sticks-with-k-sticks-visible/1866-number-of-ways-to-rearrange-sticks-with-k-sticks-visible.java
class Solution {
    static int[][] dp = new int[1001][1001];
    static int mod = (int)(1e9+7);
    public int rearrangeSticks(int n, int k) {
        if(k==0){
            return 0;
        }
        if(n==k){
            return 1;
        }
        if(dp[n][k]!=0){
            return dp[n][k];
        }
        long ans1 = rearrangeSticks(n-1,k-1);
        long ans2 = 1L * rearrangeSticks(n-1,k)*(n-1);
        ans2 %= mod;
        ans1+=ans2;
        ans1%=mod;
        return dp[n][k] = (int) (ans1);
    }
}